Bluetooth: ATT: Fix crash if bt_l2cap_send_cb fails

Vudentz opened the pull request:
This fixes a regression introduced by
10841b9a14b748a34c05cdebf0e9a278433a6ddc as it did remove a call to
net_buf_ref which was used not only to keep a reference for resending
but also to prevent bt_l2cap_send_cb to unref the buffer in case it
fails.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ATT: Fix crash if bt_l2cap_send_cb fails

This fixes a regression introduced by
10841b9a14b748a34c05cdebf0e9a278433a6ddc as it did remove a call to
net_buf_ref which was used not only to keep a reference for resending
but also to prevent bt_l2cap_send_cb to unref the buffer in case it
fails.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

joerchan commented on subsys/bluetooth/host/att.c:
```diff
@@ -189,11 +189,18 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf,
 	if (err) {
 		net_buf_simple_restore(&buf->b, &state);
+	} else {
+		net_bug_unref(buf);
```
I think you are introducing a bug here ;)

Vudentz replied:
It is releasing the reference we took to prevent hte buffer to be released while in bt_l2cap_send_cb, so what bug are you talking about here, you don't need to keep it a secrete.

joerchan replied:
You have a spelling mistake, that's the "bug"

jhedberg commented on subsys/bluetooth/host/att.c:
```diff
@@ -189,11 +189,18 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf,
 
 	chan->sent = cb ? cb : chan_cb(buf);
 
+	/* bt_l2cap_send_cb takes onwership of the buffer so take another
+	 * reference to restore the state in case an error is returned.
+	 */
+	net_buf_ref(buf);
+
 	err = bt_l2cap_send_cb(chan->att->conn, BT_L2CAP_CID_ATT,
 			       buf, att_cb(chan->sent),
 			       &chan->chan.chan);
 	if (err) {
 		net_buf_simple_restore(&buf->b, &state);
+	} else {
+		net_buf_unref(buf);
```
This all makes for very unintuitive and confusing looking code (which is true any time you have conditional unrefs), even with the code comment. Would it make sense to consider consider changing the semantics of `bt_l2cap_send_cb()` to either:

 - Always grabs a new reference to itself (i.e. caller is always responsible for unrefing)

or

 - Only take ownership in case of success. In fact, looking at the callers of `chan_send()` it looks like that's the assumption for chan send, i.e. the callers assume they still own buff when they get an error.

joerchan replied:
I think second option makes more sense. 

@Vudentz Does the same problem exist in `l2cap_chan_le_send`?
There is a `bt_l2cap_send_cb` and a `net_buf_simple_restore` there as well.

The ref handling easily leads to these kinds of mistakes, it's not the first time.

Vudentz replied:
> I think second option makes more sense.
> 
> @Vudentz Does the same problem exist in `l2cap_chan_le_send`?
> There is a `bt_l2cap_send_cb` and a `net_buf_simple_restore` there as well.
> 
> The ref handling easily leads to these kinds of mistakes, it's not the first time.

Yeah, if seg == buf it looks like there would be indeed a similar problem.

Vudentz replied:
We could perhaps introduce a version that does not unref on error, e.g. bt_conn_send_nounref so callers that want to handler the errors themselves can use that instead of the regular one that always unref, thougths?

Vudentz replied:
@jhedberg @joerchan I took the approach of changing the default behavior to not unref in case of failure since there are more code that rely on that and then introduced bt_conn_send_unref(_cb) for the code that want to stick to the old behavior.

jhedberg commented on subsys/bluetooth/host/att.c:
```diff
@@ -189,11 +189,18 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf,
 
 	chan->sent = cb ? cb : chan_cb(buf);
 
+	/* bt_l2cap_send_cb takes onwership of the buffer so take another
+	 * reference to restore the state in case an error is returned.
```
This isn't always true though, is it? If ownership was always taken you'd have an unconditional 'unref()', not a conditional one like you do now.

jhedberg replied:
Answering myself: yes, it's true. It's just that `bt_l2cap_send_cb()` and `chan_send()` have different buffer reference handling semantics, which is the cause of all this confusing code.

jhedberg commented:
This reminds me, we should probably document, both for internal and external functions, the semantics wrt buffer handling, i.e. do they take ownership and if so always or under specific conditions like success. E.g. `chan_send()` has zero documentation currently (note I don't mean doxygen documentation but just any code comment describing the behaviour of the function)

Vudentz commented:
> This reminds me, we should document, for both internal and external functions, the semantics wrt buffer handling, i.e. do they take ownership and if so always or under specific conditions like success. E.g. `chan_send()` has zero documentation currently (note I don't mean doxygen documentation but just any code comment describing the behaviour of the function)

The bt_l2cap_chan_cb do have this documented: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/l2cap_internal.h#L303

So this is specific to chan_send if does handles faiilures gracefully by restoring the state thus why it needs to keep a reference for itself, I will add a comment to chan_send regarding this logic.

Vudentz committed:
Bluetooth: ATT: Add documentation for chan_send

chan_send does restore buffer state in case of an error which is
different than how bt_l2cap_send_cb works as it does always unref in
case of an error.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

joerchan commented:
@Vudentz @jhedberg I think it would be to have the bug fixed first and then do the refactoring separately.
We could then have the fix backported, and the refactoring only on master branch.

The public API `bt_l2cap_chan_send` has now a different behavior in buffer handling.

Vudentz commented:
> @Vudentz @jhedberg I think it would be to have the bug fixed first and then do the refactoring separately.
> We could then have the fix backported, and the refactoring only on master branch.
> 
> The public API `bt_l2cap_chan_send` has now a different behavior in buffer handling.

Hmm, I didn't change the buffer management there since it would still queue the buffer or you meant to say we would need something like:

```
diff --git a/subsys/bluetooth/host/l2cap.c b/subsys/bluetooth/host/l2cap.c
index 55d59c0110..af1a084f47 100644
--- a/subsys/bluetooth/host/l2cap.c
+++ b/subsys/bluetooth/host/l2cap.c
@@ -2603,6 +2603,7 @@ int bt_l2cap_chan_send(struct bt_l2cap_chan *chan, struct net_buf *buf)
                        return data_sent(buf)->len;
                }
                BT_ERR("failed to send message %d", err);
+               net_buf_unref(buf);
        }
 
        return err;

```

Vudentz commented:
@jhedberg @joerchan Ive updated the changes, so bt_l2cap_chan_send (the only public API affected) behaves just as before.

jhedberg commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2596,6 +2603,7 @@ int bt_l2cap_chan_send(struct bt_l2cap_chan *chan, struct net_buf *buf)
 			return data_sent(buf)->len;
 		}
 		BT_ERR("failed to send message %d", err);
+		net_buf_unref(buf);
```
This doesn't look right at all. What about all the other branches of this function that may return error but do not unref the buffer?

Vudentz replied:
In those cases the buffer is always queued to be send latter (whenever there are more credits, TX context, etc).

jhedberg replied:
In what cases? I don't see any queuing happening e.g for the `-EINVAL`, `-ENOTCONN` or `-ESHUTDOWN` branches.

Vudentz replied:
Actually taking a looks at the failures case it is more common to not unref than it is, since for error like -ENOTCONN and -ESHUTDOWN we don't unref the buffer, I wonder if it would be the case of never unref even if no segments could be send which is currently the only case where the buffer would be unref which perhaps we could consider as a bug as well given that was not inline with how we treat other errors.

Vudentz replied:
@jhedberg I means in case of -ENOTCONN and -ESHUTDOWN the application would have to call net_buf_unref while other (unlikely) errors it would not have to, so Id suggested to align with the former error handling so the application always have to unref in case of error which I believe was always the intention of this API.

jhedberg replied:
Always having to unref is fine with me

Vudentz replied:
Ok, so I removed the net_buf_unref from bt_l2cap_chan_send and added a note clarifying the buffer handling to make it clearer.

Vudentz replied:
@joerchan I hope you are fine with this as well.

Vudentz commented:
@joerchan btw, I did update BR code as well.

carlescufi commented:
> @jhedberg @joerchan Ive updated the changes, so bt_l2cap_chan_send (the only public API affected) behaves just as before.

Is this still the case @Vudentz? Doesn't seem like it in the code as per inspection.

Vudentz commented:
> > @jhedberg @joerchan Ive updated the changes, so bt_l2cap_chan_send (the only public API affected) behaves just as before.
> 
> Is this still the case @Vudentz? Doesn't seem like it in the code as per inspection.

There a patch for the documentation which aligns the expect behavior since previously the way we handle errors was not consistent (most didn't unref but in the rare occasion nothing could be sent the buffer would be unref).

joerchan reviewed (changes requested):
@Vudentz Now we have two different buffer semantics for the two functions: `bt_iso_chan_send` and `bt_l2cap_chan_send`.
In my opinion we should update `bt_iso_chan_send` to have the same semantics as `bt_l2cap_chan_send` and not unref the buffer.
I think it makes for better application code to unref the buffer on error instead of ref'ing and unref on success if the application wishes to have a queue.
If we do this then we don't need the internal bt_conn_send_(cb)_unref function since there is no more users of it.

Also the buffer semantics of `bt_iso_chan_send` should be documented in the public API header.
I can see that the central_iso sample does not unref the buffer while the shell does unref it, so atleast one of them is wrong.

I guess it is a definition question for the L2CAP API if this is a behavior change or a bugfix but to me this seems like a behavior change and should be documented in the release notes.

Vudentz commented:
> @Vudentz Now we have two different buffer semantics for the two functions: `bt_iso_chan_send` and `bt_l2cap_chan_send`.
> In my opinion we should update `bt_iso_chan_send` to have the same semantics as `bt_l2cap_chan_send` and not unref the buffer.
> I think it makes for better application code to unref the buffer on error instead of ref'ing and unref on success if the application wishes to have a queue.
> If we do this then we don't need the internal bt_conn_send_(cb)_unref function since there is no more users of it.
> 
> Also the buffer semantics of `bt_iso_chan_send` should be documented in the public API header.
> I can see that the central_iso sample does not unref the buffer while the shell does unref it, so atleast one of them is wrong.
> 
> I guess it is a definition question for the L2CAP API if this is a behavior change or a bugfix but to me this seems like a behavior change and should be documented in the release notes.

Well while I agree that bt_iso_chan_send shall have the same semantics it seems the direction you are given is that this would not be backported to 2.5 leaving it broken there as well, this debate for behavior change sounds a little conflicting with the direction given initially and in the meantime we could have applied the initial fix.
